### PR TITLE
Switch the TypeScript template to a three-stage build

### PR DESCRIPTION
**Pull request by fnesveda**
Just an experiment to speed up TypeScript actor builds using three-stage Docker builds:

- 1st stage from base image to install only production dependencies
- 2nd stage which starts from 1st stage, which also installs dev dependencies, copies source files, and runs the TypeScript build
- final stage from base image, which copies the prod dependencies from 1st stage and built folder from 2nd stage

It's better over the current two-stage builds because you don't have to install the production dependencies twice (once in each stage), and you don't have to explicitly prune the npm cache.

### CODE REVIEW

**Review by mnmkng — APPROVED** [review]
Nice! Have you had a chance to try if it performs better?

File: templates/getting_started_typescript/.actor/Dockerfile [hunk 1]
```diff
@@ -1,51 +1,47 @@
...
-FROM apify/actor-node:16 AS builder
+FROM apify/actor-node:16 AS prod-dependencies
 
 # Copy just package.json and package-lock.json
 # to speed up the build using Docker layer cache.
 COPY package*.json ./
 
-# Install all dependencies. Don't audit to speed up the installation.
+# Install only prod dependencies. Don't audit to speed up the installation.
+RUN npm install --omit=dev --omit=optional --audit=false
```
[the remark this answers is not in the document]
Comment (mnmkng) [reply templates/getting_started_typescript/.actor/Dockerfile:11]: Hmm, might be missing `npm set cache ~/.npm`

[the remark this answers is not in the document]
Comment (fnesveda) [reply templates/getting_started_typescript/.actor/Dockerfile:11]: Hmm, I'm not sure how this work would work when there would be no RUN cache hit.
Right now, the NPM cache is already populated in the base image, with mostly the right packages, so it should speed up the installation quite a bit by itself. If we'd use `RUN --mount=type=cache,...` and there would be no cache for that step, would it use an empty directory for that cache (bad), or would it use whatever was on that path in the image before (good)?

It's definitely worth exploring, but I would do it separately from this optimization.

**Comment by fnesveda** [comment]
@mnmkng I did some very non-scientific testing and it felt faster, but some proper testing would be needed to see if it's actually better.

**Pull request closed without merging**